### PR TITLE
Feature/journal migrations

### PR DESCRIPTION
**Pull request by apfelfabrik**
OP Issue: https://www.openproject.org/issues/1793. To test, one can do the following:

```
bundle update                                 &&
bundle exec rake db:drop                      &&
git checkout -- Gemfile.lock                  &&
git checkout 034bcd1                          &&
bundle update                                 &&
bundle exec rake db:create db:migrate db:seed
```

Then, remove all work packages, and continue:

```
git checkout -- Gemfile.lock                  &&
git checkout feature/journal-migrations       &&
bundle update                                 &&
bundle exec rake db:migrate
```

~~:exclamation: Depends on https://github.com/opf/openproject/pull/372. No longer depends on https://github.com/opf/openproject/pull/393.~~

### CODE REVIEW

**Comment by apfelfabrik** [comment]
I have rebased and shortened this PR for easier reviews.

File: db/migrate/20130903092236_awesome_legacy_journals_migration.rb [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+      # compute the combined journal from current and all previous changesets.
+      combined_journal = legacy_journal["changed_data"]
+      if previous_journaled_id == journaled_id && previous_type == type
+        combined_journal = previous_journal.merge(combined_journal)
+      end
+
+      # remember the combined journal as the previous one for the next iteration.
+      previous_journal = combined_journal
+      previous_journaled_id = journaled_id
+      previous_type = type
```
Comment (ulferts) [inline db/migrate/20130903092236_awesome_legacy_journals_migration.rb]: You could use a struct here to semantically bundle those values.

File: db/migrate/20130903092236_awesome_legacy_journals_migration.rb [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+  end
+
+  def quoted_journals_table_name
+    @@quoted_journals_table_name ||= quote_table_name 'journals'
+  end
+
+  def check_assumptions
+
+    invalid_journals = ActiveRecord::Base.connection.select_values <<-SQL
+      SELECT DISTINCT tmp.id
```
Comment (ulferts) [inline db/migrate/20130903092236_awesome_legacy_journals_migration.rb]: Please add a comment explaining the intend of this sql

**Comment by ulferts** [comment]
When applying this migration to a current openproject.org dump some problems arise. So I would like to continue development on this before merging.

**Comment by apfelfabrik** [comment]
I also ran into broken journals and inconsistencies. I deleted some journals that wouldn't migrate, one with which I had multibyte encoding + varchar limit problem (probably due to borked setup on my side) and fixed handling of journals where no changed_data is attached (no idea where those come from, but they're in my legacy data). Also, I had to fix the not-null constraint on page_id w/ wiki content journals.

I had to change multiple unrelated migrations to get to the journal migrations, mostly stuff we either pulled into or out of the core. Those issues are in separate tickets.

I also took the liberty of adding a progressbar w/ ETA. With such long running migrations I believe that to be a good idea. We can also use it for db:seed, which takes a while, too.

**Pull request merged**